### PR TITLE
Asset server node JSON stats

### PR DESCRIPTION
**Pull request by Atlante45**
- AssetServers send stats to the DomainServer like the other nodes

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2958e400625fe38e3abb39e09eac715ffc9e6aa4 available at: https://gist.github.com/7be548904bfb858149fe<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.dmg

File: libraries/networking/src/LimitedNodeList.cpp [hunk 1]
```diff
@@ -839,6 +845,13 @@ void LimitedNodeList::sendPeerQueryToIceServer(const HifiSockAddr& iceServerSock
     sendPacketToIceServer(PacketType::ICEServerQuery, iceServerSockAddr, clientID, peerID);
 }
 
+SharedNodePointer LimitedNodeList::findNodeWithAddr(const HifiSockAddr& addr) {
+    auto it = std::find_if(std::begin(_nodeHash), std::end(_nodeHash), [&](const UUIDNodePair& pair) {
+        return pair.second->getActiveSocket() ? (*pair.second->getActiveSocket() == addr) : false;
+    });
+    return (it != std::end(_nodeHash)) ? it->second : SharedNodePointer();
```
Comment (birarda) [inline libraries/networking/src/LimitedNodeList.cpp]: pretty sure we need to guard this with the mutex

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7ded7fa8408604c01e2238977bb74d13935dbeec available at: https://gist.github.com/7be548904bfb858149fe<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.dmg

File: assignment-client/src/assets/AssetServer.cpp [hunk 2]
```diff
@@ -165,3 +163,59 @@ void AssetServer::handleAssetUpload(QSharedPointer<NLPacketList> packetList, Sha
...
+        receivingStats["ReceivedACK2"] = stat.second.events[udt::ConnectionStats::Stats::ReceivedACK2];
+        receivingStats["Duplicate"] = stat.second.events[udt::ConnectionStats::Stats::Duplicate];
+        nodeStats["receiving"] = receivingStats;
+        
+        serverStats[uuid] = nodeStats;
+    }
+    
+    // send off the stats packets
+    QJsonObject statsObject;
+    statsObject["AssetServer"] = serverStats;
```
Comment (birarda) [inline assignment-client/src/assets/AssetServer.cpp]: I don't think this namespacing is very useful - the DS will already display the stats only for each node you look at

File: domain-server/src/DomainServerNodeData.cpp [hunk 3]
```diff
@@ -10,40 +10,69 @@
...
-    _statsJSONObject = mergeJSONStatsFromNewObject(QJsonObject::fromVariantMap(packetVariantMap), _statsJSONObject);
+void DomainServerNodeData::updateJSONStats(QByteArray statsByteArray) {
+    auto document = QJsonDocument::fromBinaryData(statsByteArray);
+    Q_ASSERT(document.isObject());
+    _statsJSONObject = overrideValuesIfNeeded(document.object());
+}
+
+QJsonObject DomainServerNodeData::overrideValuesIfNeeded(const QJsonObject& newStats) {
+    QJsonObject result;
+    for(auto it = newStats.constBegin(); it != newStats.constEnd(); ++it) {
```
Comment (birarda) [inline domain-server/src/DomainServerNodeData.cpp]: space after the `for`

**Comment by hifi-gustavo** [comment]
Build console output for commit: d32689ec18cf92c1a0948f3f5b9dde3a86289371 available at: https://gist.github.com/7be548904bfb858149fe<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.dmg

File: domain-server/src/DomainServerNodeData.cpp [hunk 4]
```diff
@@ -10,40 +10,69 @@
...
+            result.push_back(value);
         }
     }
-    
-    return destinationObject;
+    return result;
+}
+
+void DomainServerNodeData::addOverrideForKey(const QString& key, const QString& value,
+                                                  const QString& overrideValue) {
```
Comment (birarda) [inline domain-server/src/DomainServerNodeData.cpp]: indentation

File: libraries/networking/src/udt/ConnectionStats.h [hunk 5]
```diff
@@ -36,11 +36,13 @@ class ConnectionStats {
...
             Retransmission,
-            Duplicate
+            Duplicate,
+            
+            NumEvent
         };
         
         // construct a vector for the events of the size of our Enum - default value is zero
-        std::vector<int> events = std::vector<int>((int) Event::Duplicate + 1, 0);
+        std::vector<int> events = std::vector<int>((int) Event::NumEvent, 0);
```
Comment (birarda) [inline libraries/networking/src/udt/ConnectionStats.h]: shouldn't this be `NumEvents`?

**Comment by hifi-gustavo** [comment]
Build console output for commit: a22904392cb895fe3b0f9b096b8d59c4febd7d65 available at: https://gist.github.com/7be548904bfb858149fe<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a06a350da17b1e4d4075348f9ce3bcf08977024c available at: https://gist.github.com/7be548904bfb858149fe<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2e1789eb128b0152cfb8e19e4e097b603427674c available at: https://gist.github.com/7be548904bfb858149fe<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 05df84ee67b4d8648ac302d6c4c9811a9143b32b available at: https://gist.github.com/7be548904bfb858149fe<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5772.dmg

**Pull request merged**